tests/fuzz/fuzz-bolt11.c

```diff
@@ -0,0 +1,138 @@
+__attribute__((weak)) size_t LLVMFuzzerMutate(uint8_t *data, size_t size,
+					      size_t max_size);
+__attribute__((weak)) size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data,
+						     size_t size,
+						     size_t max_size,
+						     unsigned int seed);
+
+void init(int *argc, char ***argv)
+{
+	chainparams = chainparams_for_network("bitcoin");
+	common_setup("fuzzer");
+}
+
+// Encodes a dummy bolt11 invoice into `fuzz_data` and returns the size of the
+// encoded string. Returns 0, if the size of the encoded dummy exceeds
+// `max_size`.
+static size_t initial_input(uint8_t *fuzz_data, size_t size, size_t max_size)
+{
+	// struct bolt11 *dummy = new_bolt11(tmpctx, NULL);
+	// dummy->chain = chainparams;
+	// char *dummy_out = bolt11_encode(tmpctx, dummy, false, &sign, NULL);
+	strcpy((char *)fuzz_data,
+	       "lnbc16lta047pp5h6lta047h6lta047h6lta047h6lta047h6lta047h6lta047"
+	       "h6lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq"
+	       "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqxnht6w");
+	size = strlen((char *)fuzz_data);
+
+	clean_tmpctx();
+	return size > max_size ? 0 : size;
+}
+
+// We use a custom mutator to produce an input corpus that consists entirely of
+// correctly encoded bech32 strings. This enables us to efficiently fuzz the
+// bolt11 decoding logic without the fuzzer getting stuck on fuzzing the bech32
+// decoding/encoding logic.
+//
+// This custom mutator does the following things:
+//   1. Attempt to bech32 decode the given input (returns the encoded dummy
+//      invoice on failure).
+//   2. Mutate either the human readable or data part of the invoice using
+//      libFuzzer's default mutator `LLVMFuzzerMutate`.
+//   3. Attempt to bech32 encode the mutated hrp and data (returns the endcoded
+//      dummy on failure).
+//   4. Write the encoded result to `fuzz_data` if its size does not exceed
+//      `max_size`, otherwise return the encoded dummy invoice.
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+			       unsigned int seed)
+{
+	if (size < 10)
+		return initial_input(fuzz_data, size, max_size);
+
+	// Interpret fuzz input as string (ensure it's null terminated).
+	char *input = to_string(tmpctx, fuzz_data, size);
+
+	// Attempt to bech32 decode the input.
+	char *hrp = tal_arr(tmpctx, char, strlen(input) - 6);
+	u5 *data = tal_arr(tmpctx, u5, strlen(input) - 8);
+	size_t datalen = 0;
+	if (bech32_decode(hrp, data, &datalen, input, (size_t)-1) !=
+	    BECH32_ENCODING_BECH32) {
+		// Decoding failed, this should only happen when starting from
+		// an empty corpus.
+		return initial_input(fuzz_data, size, max_size);
+	}
+
+	// Mutate either the hrp or data. Given the same seed, the same
+	// mutation is performed.
+	srand(seed);
+	switch (rand() % 2) {
+	case 0: // Mutate hrp and ensure it's still null terminated.
+		hrp[LLVMFuzzerMutate((uint8_t *)hrp, strlen(hrp),
+				     strlen(input) - 6) -
+		    1] = '\0';
```

Comment:
I think this will tend to chop off the last byte of the HRP each time.

For example:
```c
hrp = "lnbc";
len = strlen(hrp);  // len = 4
size_t new_len = LLVMFuzzerMutate(hrp, len, ...);  // libFuzzer mutates to hrp = "lnbd", new_len = 4
hrp[new_len - 1] = '\0';  // hrp = "lnb"
```

This might cripple libFuzzer's mutation strategy slightly.

If we pass `max_size = strlen(input) - 7` instead, we can set `hrp[new_len] = '\0'` and avoid chopping.